lib/src/chunked_stream_reader.dart

```diff
@@ -0,0 +1,176 @@
+// Copyright (c) 2021, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:async';
+import 'dart:typed_data';
+
+import 'byte_collector.dart' show collectBytes;
+
+/// Utility class for reading elements from a _chunked stream_.
+///
+/// A _chunked stream_ is a stream where each event is a chunk of elements.
+/// Byte-streams with the type `Stream<List<int>>` is common of example of this.
+/// As illustrated in the example below, this utility class makes it easy to
+/// read a _chunked stream_ using custom chunk sizes and sub-stream sizes,
+/// without managing partially read chunks.
+///
+/// ```dart
+/// final r = ChunkedStreamReader(File('myfile.txt').openRead());
+/// try {
+///   // Read the first 4 bytes
+///   final firstBytes = await r.readChunk(4);
+///   if (firstBytes.length < 4) {
+///     throw Exception('myfile.txt has less than 4 bytes');
+///   }
+///
+///   // Read next 8 kilobytes as a substream
+///   Stream<List<int>> substream = r.readStream(8 * 1024);
+///
+///   ...
+/// } finally {
+///   // We always cancel the ChunkedStreamReader, this ensures the underlying
+///   // stream is cancelled.
+///   r.cancel();
+/// }
+/// ```
+///
+/// The read-operations: [readChunk] and [readStream] cannot be invoked
+/// concurrently.
+class ChunkedStreamReader<T> {
+  final StreamIterator<List<T>> _input;
+  final List<T> _emptyList = <T>[];
```

Comment:
This list is growable. If you share it with someone else, it might not stay empty. I don't *think* you do, it's just used as a sentinel. Why not just use `null` instead to represent the empty buffer, if you are going to check for whether it's empty before using it anyway?
You're going to replace the empty _buffer with something non-empty anyway, and checking `_buffer == null` vs `_buffer.isEmpty` seems like the former should be more efficient.

Reply:
> if you are going to check for whether it's empty before using it anyway?

Well, we also handle the cases when `StreamIterator.current` is an empty list efficiently, without any additional checks.

It's nice to avoid nullable when empty-list should have the same behavior.

How about wrapping with `UnmodifiableListView` from `dart:collection`, that would also guard against mutation of `_emptyList`.

Reply:
You could make it const. `final List<T> _empty = const [];` or `final List<T> _empty = const <Never>[];`